### PR TITLE
Aurea - Edges

### PR DESCRIPTION
**Pull request by Aurea-Li**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, and how did you try querying the API? | I read the documentation and experimented around in Postman. 
What is an API Wrapper? Why is it in `lib`? How would your project change if you needed to interact with more than one API (aka more than just the Edamam API)? | An API wrapper is a way to wrap the dependency around API in one place. I'm not sure why it's in `lib`, maybe by convention? Because it is business logic but doesn't really belong to a model. If there were more than one API, there would be an API wrapper for each different API. 
Describe your API wrapper, the methods you created in it, and why you decided to create those methods/how they were helpful | My API wrapper has several methods. The two public ones are `find_recipe` and `search_recipes`. These are the ones used in controller to send messages to the API to query recipes, or find an individual recipe. 
What was an edge case or failure case test you wrote for your API Wrapper? What was a nominal case? | For `find_recipe`, the input is a valid `uri`. One of my failure cases was giving an invalid `uri`. I ensured that this would result in a `nil` output.
How does VCR aid in testing an API? | It allows you to replay API calls to minimize the number of times you are actually interacting with the API. 
What is the Heroku URL of your deployed application? | https://ayl-edamam.herokuapp.com/

### CODE REVIEW

File: app/views/recipes/results.html.erb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+<h2> Results for "<%= @query %>"</h2>
+
+
+<% if @recipes.empty? %>
+  <h3> No results found. </h3>
+<% else %>
+  <ul>
+  <% @recipes.each do |recipe| %>
+    <li>
+      <%= image_tag recipe.photo, alt: recipe.name %>
```
Comment (CheezItMan) [inline app/views/recipes/results.html.erb:13]: I suggest making the `image_tag` also a link because it's human nature to want to click on the images.

File: lib/edamam_api_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  TOKEN = ENV["RECIPE_TOKEN"]
+  TOKENS_URL = "&app_id=#{ID}&app_key=#{TOKEN}"
+
+
+  def self.search_recipes(query)
+    url = BASE_URL + "q=#{query}" + TOKENS_URL
+
+    responses = HTTParty.get(url)["hits"]
+
+    recipes = responses.map do |response|
```
Comment (CheezItMan) [inline lib/edamam_api_wrapper.rb:15]: You should have an `if` statement to ensure a valid response.

File: test/lib/edamam_api_wrapper_test.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+        expect(recipes).must_respond_to :each
+        recipes.each do |recipe|
+          expect(recipe).must_be_instance_of Recipe
+        end
+
+      end
+    end
+
+    it "returns empty list when query is unsuccessful" do
```
Comment (CheezItMan) [inline test/lib/edamam_api_wrapper_test.rb:20]: You should also test a search of the empty string or nonsense characters.

File: test/controllers/recipes_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  describe "search" do
+    it "can get search page" do
+      get root_path
+
+      must_respond_with :success
+    end
+  end
+
+  describe "results" do
+    it "can get results page" do
```
Comment (CheezItMan) [inline test/controllers/recipes_controller_test.rb:21]: You should also include a negative test-case for the search results and show actions.

File: app/controllers/recipes_controller.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+
+  def search
+  end
+
+  def results
+
+    @query = params[:query]
+
+    @recipes = EdamamApiWrapper.search_recipes(@query).paginate(page: params[:page], per_page: 10)
```
Comment (CheezItMan) [inline app/controllers/recipes_controller.rb:13]: I believe because your API is only pulling in 10 results it's not paginating.

File: lib/edamam_api_wrapper.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
+require 'httparty'
+
+class EdamamApiWrapper
+  BASE_URL = "https://api.edamam.com/search?"
+  ID = ENV["RECIPE_ID"]
+  TOKEN = ENV["RECIPE_TOKEN"]
+  TOKENS_URL = "&app_id=#{ID}&app_key=#{TOKEN}"
```
Comment (CheezItMan) [inline lib/edamam_api_wrapper.rb:7]: A URL like this will only give you 10 responses, if you include `&to=100` you'll get 100 responses.

**Comment by CheezItMan** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene |  Good number of commits and good commit messages
Comprehension questions |  Check, The wrapper code goes into the `lib` folder because it is a class which doesn't fall into the typical Rails mode, view or controller.  
**General** |
Rails fundamentals (RESTful routing, use of named paths) |  RESTful enough, you could make the `/results` route, be `/recipes` since it display a list of recipes.  
Semantic HTML |  Great semantic HTML
Errors are reported to the user |  For most things, but if I search for `&*^` the app crashes
API Wrapper to handle the API requests |  Check, see my notes on the search action.  
Controller testing |  Check, but you should have **negative** cases.  
Lib testing |  Check, but see my notes
Search Functionality |  Check
List Functionality |  Check
Show individual item functionality |  Check
**Styling** |
List view shows 10 items at a time and/or has pagination |  **missing pagination**
Attractive and usable UI |  Nice, good responsive design!
**API Features** |
The App attributes Edamam |  Check
The VCR cassettes do not contain the API key |  Check
**External Resources** |
Link to deployed app on Heroku |  Check
**Overall** |  Pretty good, see my notes on your testing and the pagination.  You hit the learning goals, but could use improvement on testing negative cases.
